File: calendar/make-cal.py
#!/usr/bin/env python3

# Reads the calendar .rst files and produces corresponding .html
#
# Python is not a beautiful language. However, it's too easily available...

from bs4 import BeautifulSoup
from docutils import core

print('Running make-cal.py...')

categories = {
    'food': ['śniadanie', 'obiad', 'kolacja', 'ognisko'],
    'play': ['poznawczy', 'memoriał', 'maraton', 'mpp', 'blitz',
             '9×9', '13×13', 'torus', 'rengo', 'fantom'],
    'learn': ['zajęcia', 'tsumego', 'wykład', 'podział', 'relay'],
}

def find_category(event):
    if ":00" in event:
        # uhh skip the hours...
        return ''

    for category in categories:
        for keyword in categories[category]:
            if keyword in event.lower():
                return category
    return 'other'

def process_rst(src):
    document = core.publish_parts(src, writer_name='html5')
    soup = BeautifulSoup(document['body'], 'html.parser')

    # add .cal class to the table
    for table in soup.find_all('table'):
        table['class'] = 'cal'

    # remove colgroup messing up layout...
    for colgroup in soup.find_all('colgroup'):
        colgroup.decompose()

    # strip p from headers
    for th in soup.find_all('th'):
        p = th.find('p')
        if p:
            th.contents = p.contents

    # process table fields: strip p, add class info
    for td in soup.find_all('td'):
        p = td.find('p')
        if p:
            event = p.contents[0].replace('\n', ' ')
            p.contents[0].replace_with(event)

            td.contents = p.contents
            td['class'] = find_category(event)

    return soup


files = ['cal-1', 'cal-2']

for f in files:
    print(f'Calendar: starting processing {f}...')

    src = open(f'./calendar/{f}.rst').read()
    soup = process_rst(src)

    with open(f'./_includes/{f}.html', 'w') as dst:
        dst.write(str(soup))
        print(f'Calendar: finished processing {f}!')

File: _plugins/make_cal.rb
module Jekyll
  class MakeCal < Generator
    # Perhaps consider sleeping 22-7 ?
    @@hrs = [
      "08", "09", "10", "11", "12", "13", "14",
      "15", "16", "17", "18", "19", "20", "21",
    ]

    # How the source should mark nothing happening
    @@empty = "~"

    @@categories = {
      food: ['śniadanie', 'obiad', 'kolacja', 'ognisko'],
      play: ['poznawczy', 'memoriał', 'maraton', 'mpp', 'blitz',
             '9×9', '13×13', 'torus', 'rengo', 'fantom', 'hex go'],
      learn: ['zajęcia', 'tsumego', 'wykład', 'podział', 'relay'],
    }

    # Parse schedule for a single day
    def parse_day(day_description)
      day, hours = day_description.split(/\n/, 2)
      hour_list = hours.strip.split(/\n/).map do |hour|
        match = hour.strip.match(/(\d+) (.*)/)
        { match[1] => match[2] }
      end
      { day: day, hours: hour_list.reduce({}, :merge) }
    end

    # Make sure the missing hours are filled properly
    def hours_expand(hours)
      new = {}
      running_event = @@empty
      @@hrs.each do |hr|
        if hours.include?(hr)
          running_event = hours[hr]
        end
        new[hr] = running_event
      end
      new
    end

    # Flip from [{08: event, 09: another}, {08: event, 09: breakfast}]
    # to {8: [event, event], 9: [another, breakfast]}
    def flip_hours(days)
      hours = {}
      days_expanded = days.map { |day| hours_expand(day) }
      days_expanded.each do |hr|
        hr.each do |key, value|
          hours[key.to_sym] ||= []
          hours[key.to_sym] << value
        end
      end
      hours
    end

    # Vertical merging: merge same events in the same column
    def add_rowspan(basic_table)
      columns = basic_table.transpose

      merged_columns = columns.map.with_index do |col, offset|
        # Build a new column array by grouping adjacent cells with the same event
        new_col = Array.new(col.size)
        i = 0
        while i < col.size
          cell = col[i]
          rowspan = col.drop(i).take_while { |c| c[:event] == cell[:event] }.size
          new_col[i] = cell.merge(rowspan: rowspan, colspan: 1, offset: offset)
          i += rowspan
        end
        new_col
      end
      # Transpose back to rows and remove nil cells per row
      merged_columns.transpose.map { |row| row.compact }
    end

    # Horizontal merging: only merge cells if they are immediately adjacent
    def add_colspan(row)
      merged = []
      current_cell = nil
      row.each do |cell|
        if current_cell.nil?
          current_cell = cell
        else
          # Only merge if the cell is immediately adjacent (i.e. its original offset equals
          # the current cell's offset plus its colspan), and if event and rowspan match.
          if cell[:event] == current_cell[:event] &&
             cell[:rowspan] == current_cell[:rowspan] &&
             (current_cell[:offset] + current_cell[:colspan] == cell[:offset])
            current_cell = current_cell.merge(colspan: current_cell[:colspan] + cell[:colspan])
          else
            merged << current_cell
            current_cell = cell
          end
        end
      end
      merged << current_cell if current_cell
      merged
    end

    # Find category for event, processed sequentially, first match
    def find_category(event)
      @@categories.each do |category, keywords|
        keywords.each do |keyword|
          return category if event.downcase.include?(keyword)
        end
      end
      'other'
    end

    # Various single-cell enhancements, last thing we do
    def preprocess_cell(cell)
      if cell[:event] == @@empty
        cell[:event] = "" # nothing!
      elsif cell[:event] =~ /^\d\d$/
        cell[:event] = cell[:event] + ":00"
      else
        cell[:class] = find_category(cell[:event])
      end
      cell
    end

    def make_table(content)
      week, days_str = content.split(/\n/, 2)
      week = week.sub("# ", "")
      days_split = days_str.strip.split(/## /)
      days_split.shift # remove empty first item

      days = days_split.map { |day_description| parse_day(day_description) }
      by_hours = flip_hours(days.map { |h| h[:hours] })

      basic_table = by_hours.values.map { |hour| hour.map { |event| { event: event } } }
      rowspanned = add_rowspan(basic_table)
      colspanned = rowspanned.map { |row| add_colspan(row) }

      # Prepend each row with an hour label cell.
      with_hours = []
      @@hrs.each_with_index do |hr, index|
        with_hours << colspanned[index].unshift({ event: hr, colspan: 1, rowspan: 1, offset: 0 })
      end

      hours = with_hours.map { |row| row.map { |cell| preprocess_cell(cell) } }

      template = File.read('_plugins/cal_template.html.erb')
      ERB.new(template).result(binding)
    end

    def generate(site)
      files = ["cal-1", "cal-2"]
      files.each do |fid|
        content = File.read("calendar/#{fid}.md")
        File.write("_includes/#{fid}.html", make_table(content))
      end
    end
  end
end
